In [4]:
from cmlreaders import CMLReader, get_data_index
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import scipy
from scipy import stats
pd.set_option('display.max_columns', None)

In [2]:
BAI_high = 15
BAI_low = 8
BDI_high = 16
BDI_low = 10

sel_sub = pd.DataFrame([])
master = pd.read_excel(r'./220302_all_neuropsych.xlsx',engine="openpyxl") #read master spreadsheet of all RAM subjects demo/neuropsych data

sub_mat=list(np.zeros(len(master))*np.nan)
df = get_data_index('r1')
# df_select = pd.DataFrame([])
for sub in df['subject'].unique():
    num = int(sub[2:5])
    for i in master.index:
        sel = master.iloc[i]['Subject Number']
        if sel == num:
            sub_mat[i]=sub
# df_select[0:50]
master['subject']=sub_mat

len(master)

601

In [5]:
master

Subject Number                  Study site  Gender  Years of Education  \
0                 1  University of Pennsylvania  Female                13.0   
1                 2  University of Pennsylvania  Female                16.0   
2                 3  University of Pennsylvania  Female                13.0   
3                 4        Dartmouth University  Female                12.0   
4                 5  University of Pennsylvania  Female                 NaN   
..              ...                         ...     ...                 ...   
596             597          Jefferson Hospital    Male                 NaN   
597             598            Emory University  Female                12.0   
598             599             UT Southwestern    Male                 NaN   
599             600             UT Southwestern    Male                15.0   
600             601            Emory University     NaN                 NaN   

     Age of seizure onset  Age at Implant (calculated) Phase II Implant Date  \
0                    20.0                    48.105026            2014-10-08   
1                    34.0                    49.383743            2014-11-12   
2                    37.0                    39.390382            2014-11-13   
3                     6.0                    51.918931            2014-11-06   
4                     NaN                    20.304432            2015-01-14   
..                    ...                          ...                   ...   
596                   NaN                          NaN            2021-12-10   
597                  16.0                    39.568346            2021-12-15   
598                   8.0                    20.728808            2022-02-04   
599                   6.0                    27.896534            2018-03-27   
600                   NaN                          NaN                   NaT   

    Phase III Implant Date   \
0                       NaT   
1                       NaT   
2                       NaT   
3                       NaT   
4                       NaT   
..                      ...   
596                     NaT   
597                     NaT   
598                     NaT   
599              2022-02-07   
600                     NaT   

                    Traumatic Brain Injury (TBI) Notes  \
0                                                  NaN   
1                                                  NaN   
2    motor vehicle accident due to loss of consciou...   
3    In early 90s, motor vehicle accident resulting...   
4                                                  NaN   
..                                                 ...   
596                                                NaN   
597                                                NaN   
598                                                NaN   
599                                                NaN   
600                                                NaN   

    Date of Pre-operative Report   BAI   BDI  PHQ-9  GAD-7 FSIQ4  FSIQ3  \
0                            NaT   8.0   2.0    NaN    NaN    72    NaN   
1                            NaT   9.0   0.0    NaN    NaN    86    NaN   
2                            NaT   NaN   NaN    NaN    NaN   102    NaN   
3                            NaT   NaN   NaN    NaN    NaN    72    NaN   
4                            NaT   NaN   NaN    NaN    NaN    80    NaN   
..                           ...   ...   ...    ...    ...   ...    ...   
596                          NaT   NaN   NaN    NaN    NaN   NaN    NaN   
597                   2019-02-19  35.0  25.0    NaN    NaN    86    NaN   
598                   2019-06-03   NaN   NaN    NaN    NaN    87    NaN   
599                   2017-06-27   NaN   2.0    NaN    NaN   NaN    NaN   
600                          NaT   NaN   NaN    NaN    NaN   NaN    NaN   

     FSIQ2 (Four Subtest)  FSIQ2 (Two Subtest)  FSIQ1 (Four Subtest)  \
0                     NaN                  NaN      

In [25]:
iq = []
iq_type = []
dep = []
dep_type = []
anx = []
anx_type = []

for i in master.index:
    try:
        if pd.notnull(master.iloc[i]['FSIQ4']) == True:
            IQ = master.iloc[i]['FSIQ4']
            IQ_type = "FSIQ4"
        elif pd.notnull(master.iloc[i]['FSIQ3']) == True:
            IQ = master.iloc[i]['FSIQ3']
            IQ_type = "FSIQ3"
        elif pd.notnull(master.iloc[i]['FSIQ2 (Four Subtest)']) == True:
            IQ = master.iloc[i]['FSIQ2 (Four Subtest)']
            IQ_type = "4FSIQ2"
        elif pd.notnull(master.iloc[i]['FSIQ2 (Two Subtest)']) == True:
            IQ = master.iloc[i]['FSIQ2 (Two Subtest)']
            IQ_type = "2FSIQ2"
        elif pd.notnull(master.iloc[i]['FSIQ1 (Four Subtest)']) == True:
            IQ = master.iloc[i]['FSIQ1 (Four Subtest)']
            IQ_type = "4FSIQ1"
        elif pd.notnull(master.iloc[i]['FSIQ1 (Two Subtest)']) == True:
            IQ = master.iloc[i]['FSIQ1 (Two Subtest)']
            IQ_type = "2FSIQ1"
        elif pd.notnull(master.iloc[i]['GAI']) == True:
            IQ = master.iloc[i]['GAI']
            IQ_type = "GAI"
        elif pd.notnull(master.iloc[i]['VCI']) == True:
            IQ = master.iloc[i]['VCI']
            IQ_type = "VCI"
        elif pd.notnull(master.iloc[i]['PRI']) == True:
            IQ = master.iloc[i]['PRI']
            IQ_type = "PRI"
        elif pd.notnull(master.iloc[i]['WMI']) == True:
            IQ = master.iloc[i]['WMI']
            IQ_type = "WMI"
        else:
            IQ = np.nan
            IQ_type = np.nan
        iq.append(IQ)
        iq_type.append(IQ_type)

        if pd.notnull(master.iloc[i]['BDI']) == True:
            Dep = master.iloc[i]['BDI']
            Dep_type = "BDI"
#         elif pd.notnull(master.iloc[i]['PHQ-9']) == True:   ### Need to normalize PHQ-9 and GAD-7 before consolidating these
#             dep = master.iloc[i]['PHQ-9']
#             Dep_type = "PHQ-9"
        else:
            Dep = np.nan
            Dep_type = np.nan
        dep.append(Dep)
        dep_type.append(Dep_type)
        
        if pd.notnull(master.iloc[i]['BAI']) == True:
            Anx = master.iloc[i]['BAI']
            Anx_type = "BAI"
#         elif pd.notnull(master.iloc[i]['GAD-7']) == True:
#             anx = master.iloc[i]['GAD-7']
#             Anx_type = "GAD-7"
        else:
            Anx = np.nan
            Anx_type = np.nan
        anx.append(Anx)
        anx_type.append(Anx_type)

    except:
        print(i)

master["IQ type"] = iq_type
master["dep type"] = dep_type
master["anx type"] = anx_type
master['IQ'] = iq
master['depression'] = dep
master['anxiety'] = anx
#master[master.depression.apply(lambda x: type(x)) != int]
master

Subject Number                  Study site  Gender  Years of Education  \
0                 1  University of Pennsylvania  Female                13.0   
1                 2  University of Pennsylvania  Female                16.0   
2                 3  University of Pennsylvania  Female                13.0   
3                 4        Dartmouth University  Female                12.0   
4                 5  University of Pennsylvania  Female                 NaN   
..              ...                         ...     ...                 ...   
596             597          Jefferson Hospital    Male                 NaN   
597             598            Emory University  Female                12.0   
598             599             UT Southwestern    Male                 NaN   
599             600             UT Southwestern    Male                15.0   
600             601            Emory University     NaN                 NaN   

     Age of seizure onset  Age at Implant (calculated) Phase II Implant Date  \
0                    20.0                    48.105026            2014-10-08   
1                    34.0                    49.383743            2014-11-12   
2                    37.0                    39.390382            2014-11-13   
3                     6.0                    51.918931            2014-11-06   
4                     NaN                    20.304432            2015-01-14   
..                    ...                          ...                   ...   
596                   NaN                          NaN            2021-12-10   
597                  16.0                    39.568346            2021-12-15   
598                   8.0                    20.728808            2022-02-04   
599                   6.0                    27.896534            2018-03-27   
600                   NaN                          NaN                   NaT   

    Phase III Implant Date   \
0                       NaT   
1                       NaT   
2                       NaT   
3                       NaT   
4                       NaT   
..                      ...   
596                     NaT   
597                     NaT   
598                     NaT   
599              2022-02-07   
600                     NaT   

                    Traumatic Brain Injury (TBI) Notes  \
0                                                  NaN   
1                                                  NaN   
2    motor vehicle accident due to loss of consciou...   
3    In early 90s, motor vehicle accident resulting...   
4                                                  NaN   
..                                                 ...   
596                                                NaN   
597                                                NaN   
598                                                NaN   
599                                                NaN   
600                                                NaN   

    Date of Pre-operative Report   BAI   BDI  PHQ-9  GAD-7 FSIQ4  FSIQ3  \
0                            NaT   8.0   2.0    NaN    NaN    72    NaN   
1                            NaT   9.0   0.0    NaN    NaN    86    NaN   
2                            NaT   NaN   NaN    NaN    NaN   102    NaN   
3                            NaT   NaN   NaN    NaN    NaN    72    NaN   
4                            NaT   NaN   NaN    NaN    NaN    80    NaN   
..                           ...   ...   ...    ...    ...   ...    ...   
596                          NaT   NaN   NaN    NaN    NaN   NaN    NaN   
597                   2019-02-19  35.0  25.0    NaN    NaN    86    NaN   
598                   2019-06-03   NaN   NaN    NaN    NaN    87    NaN   
599                   2017-06-27   NaN   2.0    NaN    NaN   NaN    NaN   
600                          NaT   NaN   NaN    NaN    NaN   NaN    NaN   

     FSIQ2 (Four Subtest)  FSIQ2 (Two Subtest)  FSIQ1 (Four Subtest)  \
0                     NaN                  NaN      

In [26]:
site_stats = pd.DataFrame([])
tot = 0

#Check how many subjects have all measures out of those that were tested and have a depression measure
for site in master['Study site'].unique():
    site_df = master[master['Study site']==site]
#     tested = site_df.dropna(subset=['subject', 'depression'])
    tested = site_df.dropna(subset=['subject'])
    tot = tot + len(tested)
    site_stats=site_stats.append({
        'site':site,
        'total subs':len(site_df),
        'subs tested':len(tested),
        'gender':tested['Gender'].count(),
        'years of edu':tested['Years of Education'].count(),
        'age of sz onset':tested['Age of seizure onset'].count(),
        'age at implant':tested['Age at Implant (calculated)'].count(),
        'TBI':tested['Traumatic Brain Injury (TBI) Notes'].count(),
        'BAI':tested['BAI'].count(),
        'BDI':tested['BDI'].count(),
        'PHQ-9':tested['PHQ-9'].count(),
        'GAD-7':tested['GAD-7'].count(),
        'FSIQ4':tested['FSIQ4'].count(),
        'FSIQ3':tested['FSIQ3'].count(),
        '4FSIQ2':tested['FSIQ2 (Four Subtest)'].count(),
        '2FSIQ2':tested['FSIQ2 (Four Subtest)'].count(),
        '4FSIQ1':tested['FSIQ1 (Four Subtest)'].count(),
        '2FSIQ1':tested['FSIQ1 (Four Subtest)'].count(),
        'GAI':tested['GAI'].count(),
        'VCI':tested['VCI'].count(),
        'PRI':tested['PRI'].count(),
        'WMI':tested['WMI'].count(),
        'Consolidated IQ':tested['IQ'].count(),
        'Consolidated dep':tested['depression'].count(),
        'Consolidated anx':tested['anxiety'].count(),
        'Neurosurg procedure':tested['Prior Neurosurgical Procedures (can select more than one)'].count()
    }, ignore_index=True)
    
site_stats[['Consolidated IQ', 'Consolidated dep', 'Consolidated anx', 'site', 'subs tested']]
#site_stats

Consolidated IQ  Consolidated dep  Consolidated anx  \
0             27.0              20.0               9.0   
1             44.0              41.0              33.0   
2             68.0              74.0              73.0   
3              2.0               0.0               0.0   
4             79.0              40.0              38.0   
5             46.0               3.0               0.0   
6             70.0              12.0               6.0   
7             19.0               1.0               1.0   
8             23.0               6.0               3.0   

                         site  subs tested  
0  University of Pennsylvania         31.0  
1        Dartmouth University         52.0  
2          Jefferson Hospital         79.0  
3    University of Washington          2.0  
4            Emory University         82.0  
5                 Mayo Clinic         58.0  
6             UT Southwestern        107.0  
7         Columbia University         27.0  
8                       NINDS         31.0

In [27]:
master.columns
# subset = master.dropna(subset=['subject'])
subset = master.dropna(subset=['IQ', 'depression', 'anxiety', 'Years of Education', 'Age at Implant (calculated)', 'Gender'])
print("Number of subjects with all essential demographic/neuropsych measures = " + str(len(subset)))

Number of subjects with all essential demographic/neuropsych measures = 148


In [13]:
trial_df = pd.DataFrame([])
list_df = pd.DataFrame([])

experiments = ['FR1', 'catFR1']

for sub in master['subject']:
#   try:
    for exp in experiments:
        localizations=df[(df['subject']==sub) & (df['experiment']==exp)]['localization'].unique()        

        for loc in localizations:
            montages = df[(df['subject']==sub) & (df['experiment']==exp) & (df['localization']==loc)]['montage'].unique()            

            for mont in montages: # iterate over montages            
                data = df[np.logical_and(df["subject"] == sub, df['experiment']==exp)]
                sessions=data[np.logical_and(data['montage']==mont,data['localization']==loc)]['session'].unique()

                for sess in sessions: 
                    reader = CMLReader(sub, exp, sess, montage=mont, localization=loc)   #reader for loading CML data
                    evs = reader.load('task_events')
                    word_evs = evs[evs['type']=='REC_WORD']

                    #exclude practice lists 
                    lists = len(word_evs['list'][word_evs['list']>0].unique()) 
                    corr_recs = len(word_evs[word_evs['intrusion']==0])
                    ELIs = len(word_evs[word_evs['intrusion']==-1])
                    PLIs = len(word_evs[word_evs['intrusion']>0])
                    total_ints = ELIs + PLIs
                    total_trials = corr_recs + total_ints

                    trials = [] #make list of event numbers of analysis
                    for list_num in word_evs['list'][word_evs['list']>0].unique():
                        list_events=word_evs[word_evs['list']==list_num]
                        if len(list_events)>0: 
                            for i in range(0,len(list_events)):
                                trials.append(list_events.index[i])
#                         PLI = len(list_events[list_events['intrusion']>0])
#                         ELI = len(list_events[list_events['intrusion']==-1])
#                         correc = len(list_events[list_events['intrusion']==0])
#                         list_df = list_df.append({ #create dataframe of events for analysis grouped by list ##should I filter out lists with no events?
#                             'subject':sub,
#                             'session': sess,
#                             'experiment': exp,
#                             'list': list_num,
#                             'number of correct recalls': correc,
#                             'number of extra-list intrusions': ELI,
#                             'number of prior-list intrusions': PLI,
#                             'total number of intrusions': ELI+PLI,
#                             'BAI':BAI,
#                             'BDI':BDI
#                             },ignore_index=True)
#                     row = master[master['subject']==sub] #get row of BAI/BDI scores for subject
#                     BAI = row['BAI'].item()
#                     BDI = row['BDI'].item()
                    trial_df = trial_df.append({ #create dataframe of events for analysis grouped by session
                        'subject':sub,
                        'session': sess,
                        'montage': mont,
                        'localization': loc,
                        'experiment': exp,
                        'number of lists': lists,
                        'number of correct recalls': corr_recs,
                        'number of extra-list intrusions': ELIs,
                        'number of prior-list intrusions': PLIs,
                        'total number of intrusions': total_ints,
                        'total number of trials left': total_trials,
                        'trials for analysis': trials
                        },ignore_index=True) 
#     except:
#         print('Failed for subject:', sub)
trial_df[:100]

experiment  localization  montage  number of correct recalls  \
0         FR1           0.0      0.0                       62.0   
1         FR1           0.0      0.0                       63.0   
2         FR1           0.0      0.0                      117.0   
3         FR1           0.0      0.0                      120.0   
4         FR1           0.0      0.0                       86.0   
..        ...           ...      ...                        ...   
95        FR1           0.0      0.0                       45.0   
96        FR1           0.0      0.0                      102.0   
97        FR1           0.0      0.0                      109.0   
98     catFR1           0.0      0.0                       99.0   
99     catFR1           0.0      0.0                       17.0   

    number of extra-list intrusions  number of lists  \
0                              21.0             22.0   
1                              19.0             23.0   
2                               3.0             25.0   
3                               2.0             25.0   
4                              30.0             22.0   
..                              ...              ...   
95                              6.0             11.0   
96                             13.0             24.0   
97                             11.0             24.0   
98                             22.0             25.0   
99                             10.0              8.0   

    number of prior-list intrusions  session subject  \
0                              25.0      0.0  R1001P   
1                              17.0      1.0  R1001P   
2                               8.0      0.0  R1002P   
3                               3.0      1.0  R1002P   
4                              26.0      0.0  R1003P   
..                              ...      ...     ...   
95                              2.0      1.0  R1066P   
96                             14.0      2.0  R1066P   
97                              8.0      3.0  R1066P   
98                             13.0      0.0  R1066P   
99                              1.0      1.0  R1066P   

    total number of intrusions  total number of trials left  \
0                         46.0                        108.0   
1                         36.0                         99.0   
2                         11.0                        128.0   
3                          5.0                        125.0   
4                         56.0                        142.0   
..                         ...                          ...   
95                         8.0                         53.0   
96                        27.0                        129.0   
97                        19.0                        128.0   
98                        35.0                        134.0   
99                        11.0                         28.0   

                                  trials for analysis  
0   [63, 64, 66, 67, 88, 89, 90, 91, 92, 93, 95, 9...  
1   [38, 39, 40, 41, 62, 63, 64, 65, 66, 67, 68, 8...  
2   [38, 39, 40, 41, 67, 68, 69, 70, 71, 74, 95, 9...  
3   [38, 39, 40, 41, 42, 43, 44, 65, 66, 67, 68, 6...  
4   [38, 39, 40, 41, 42, 44, 46, 48, 69, 70, 71, 7...  
..                                                ...  
95  [38, 39, 41, 42, 63, 64, 65, 66, 67, 68, 69, 9...  
96  [39, 40, 41, 42, 43, 44, 45, 46, 67, 68, 69, 7...  
97  [39, 40, 41, 42, 43, 44, 66, 67, 69, 70, 92, 9...  
98  [39, 40, 41, 43, 44, 45, 46, 69, 70, 72, 73, 7...  
99  [82, 83, 104, 107, 108, 109, 110, 111, 133, 13...  

[100 rows x 12 columns]

In [50]:
# master[master['subject']=='R1001P']['Phase III Implant Date'].item()
# master.columns
np.isnan(master.iloc[0]['Prior Neurosurgical Procedures (can select more than one)'])

True

In [56]:
#BEGINNING OF SESSION LEVEL ANALYSES (?)
df_master = pd.DataFrame([])
#Create master dataframe of important session stats, demographics, and neuropsych.
for i in trial_df.index:
    try:
        sub = trial_df.iloc[i]['subject']
        eli = trial_df.iloc[i]['number of extra-list intrusions']
        pli = trial_df.iloc[i]['number of prior-list intrusions']
        correc = trial_df.iloc[i]['number of correct recalls']
        allrec = trial_df.iloc[i]['total number of trials left']
        lists = trial_df.iloc[i]['number of lists']
        session=trial_df.iloc[i]['session']
        exp = trial_df.iloc[i]['experiment']
        PLI_avg = float(pli/lists) #average per list
        ELI_avg = float(eli/lists)
        correc_avg = float(correc/lists)
        allrec_avg = float(allrec/lists)
#         neuro = master[master['subject']==sub]['Prior Neurosurgical Procedures (can select more than one)'].item()
#         if np.isnan(neuro)==False:
#             binary_neuro = 1
#         elif np.isnan(neuro)==True:
#             binary_neuro = 0
        if allrec > 0: #remove sessions with no valid recalls and no BAI or BDI
            df_master = df_master.append({
                'subject':sub,
                'experiment':exp,
                'correct recalls':correc,
                'ELIs':eli,
                'PLIs':pli,
                'total ints': eli+pli,
                'lists':lists,
                'session':session,
                'all recalls':allrec,
                'average_correct_recalls':correc_avg,
                'average_total_recalls':allrec_avg,
                'average_ELI':ELI_avg,
                'average_PLI':PLI_avg,
                'average_total_ints':ELI_avg+PLI_avg,
                'BAI':master[master['subject']==sub]['BAI'].item(),
                'BDI':master[master['subject']==sub]['BDI'].item(),
                'site':master[master['subject']==sub]['Study site'].item(),
                'gender':master[master['subject']==sub]['Gender'].item(),
                'years_of_education':master[master['subject']==sub]['Years of Education'].item(),
                'age_of_seizure_onset':master[master['subject']==sub]['Age of seizure onset'].item(),
                'age_at_implant':master[master['subject']==sub]['Age at Implant (calculated)'].item(),
                'implant_date':master[master['subject']==sub]['Phase II Implant Date'].item(),
                'reimplant_date':master[master['subject']==sub]['Phase III Implant Date '].item(),
                'TBI_notes':master[master['subject']==sub]['Traumatic Brain Injury (TBI) Notes'].item(),
                'binary_TBI':pd.isnull(master[master['subject']==sub]['Traumatic Brain Injury (TBI) Notes'].item()),
                'date_of_neuropsych_report':master[master['subject']==sub]['Date of Pre-operative Report'].item(),
                'IQ':master[master['subject']==sub]['IQ'].item(),
                'depression':master[master['subject']==sub]['depression'].item(),
                'anxiety':master[master['subject']==sub]['anxiety'].item(),
                'IQ_type':master[master['subject']==sub]['IQ type'].item(),
                'dep_type':master[master['subject']==sub]['dep type'].item(),
                'anx_type':master[master['subject']==sub]['anx type'].item(),
#                 'binary_prior_neuro':binary_neuro,
                'binary_prior_neuro': pd.isnull(master[master['subject']==sub]['Prior Neurosurgical Procedures (can select more than one)'].item()), 
                'prior_neurosurgical_procedures':master[master['subject']==sub]['Prior Neurosurgical Procedures (can select more than one)'].item(),    
            }, ignore_index = True)
        else:
            print('removed row =', i, sub, exp, 'due to no valid recalls')
    except:
        print('failed for row =', i, sub, exp)
# df_master = df_master.dropna(subset=['BAI', 'BDI'], how='all') #drop sessions from subjects that have neither score
df_master[:50]


/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:14: RuntimeWarning: invalid value encountered in double_scalars
  
/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:15: RuntimeWarning: invalid value encountered in double_scalars
  from ipykernel import kernelapp as app
/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:16: RuntimeWarning: invalid value encountered in double_scalars
  app.launch_new_instance()
/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:17: RuntimeWarning: invalid value encountered in double_scalars


removed row = 12 R1015J FR1 due to no valid recalls
removed row = 158 R1100D FR1 due to no valid recalls
removed row = 159 R1100D FR1 due to no valid recalls


/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:16: RuntimeWarning: divide by zero encountered in double_scalars
  app.launch_new_instance()
/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/ipykernel_launcher.py:17: RuntimeWarning: divide by zero encountered in double_scalars


BAI   BDI  ELIs     IQ IQ_type  PLIs  \
0    8.0   2.0  21.0   72.0   FSIQ4  25.0   
1    8.0   2.0  19.0   72.0   FSIQ4  17.0   
2    9.0   0.0   3.0   86.0   FSIQ4   8.0   
3    9.0   0.0   2.0   86.0   FSIQ4   3.0   
4    NaN   NaN  30.0  102.0   FSIQ4  26.0   
5    NaN   NaN  27.0  102.0   FSIQ4  47.0   
6    NaN   NaN   3.0   72.0   FSIQ4   6.0   
7    7.0   7.0   9.0   83.0   FSIQ4  21.0   
8    7.0   7.0   7.0   83.0   FSIQ4   3.0   
9    2.0   1.0   2.0   92.0     GAI  12.0   
10   2.0   1.0   4.0   92.0     GAI   6.0   
11   2.0  12.0   0.0   98.0   FSIQ4   4.0   
12   7.0  15.0   1.0   89.0     GAI   3.0   
13   NaN  11.0  14.0   76.0   FSIQ3  11.0   
14   NaN  11.0  17.0   76.0   FSIQ3  13.0   
15   0.0   2.0   2.0   75.0   FSIQ4  15.0   
16   0.0   4.0   2.0   98.0     GAI   7.0   
17   NaN   0.0  22.0   70.0   FSIQ4   8.0   
18   3.0   1.0  25.0   74.0     GAI  14.0   
19   4.0   2.0   1.0   92.0   FSIQ3   1.0   
20   2.0  12.0   1.0   98.0   FSIQ4   5.0   
21   NaN  35.0   5.0  106.0   FSIQ4   5.0   
22   NaN  35.0   5.0  106.0   FSIQ4   3.0   
23  33.0  19.0  14.0   89.0     GAI   8.0   
24   NaN   NaN   3.0  103.0  4FSIQ1   6.0   
25   NaN   NaN   6.0  103.0  4FSIQ1  18.0   
26  28.0  11.0  15.0  100.0     GAI  13.0   
27   NaN   NaN   0.0   88.0  4FSIQ1   3.0   
28   NaN   NaN   1.0   88.0  4FSIQ1   6.0   
29   3.0   6.0   3.0   99.0   FSIQ4   7.0   
30   3.0   6.0   2.0   99.0   FSIQ4   2.0   
31   3.0   6.0   2.0   99.0   FSIQ4   1.0   
32   4.0   3.0   4.0   73.0   FSIQ4   4.0   
33   NaN  10.0   2.0   84.0   FSIQ4   3.0   
34   NaN  10.0  17.0   84.0   FSIQ4  11.0   
35   NaN  10.0   1.0   84.0   FSIQ4   1.0   
36   NaN  10.0   3.0   78.0   FSIQ3   3.0   
37   NaN  10.0  10.0   78.0   FSIQ3  10.0   
38   NaN   NaN  12.0   84.0  4FSIQ1  30.0   
39   NaN   NaN  11.0   84.0  4FSIQ1  20.0   
40   NaN   NaN  24.0   80.0  4FSIQ2  22.0   
41   NaN   NaN  18.0   80.0  4FSIQ2  11.0   
42   NaN   NaN   1.0   70.0  4FSIQ1   3.0   
43   NaN   NaN   1.0  119.0  4FSIQ1   1.0   
44   NaN   NaN   0.0  119.0  4FSIQ1   3.0   
45   2.0   7.0   8.0  111.0     GAI   2.0   
46   2.0   7.0   0.0  111.0     GAI   0.0   
47   NaN   NaN  10.0  113.0   FSIQ4  33.0   
48   NaN   NaN   6.0  113.0   FSIQ4   8.0   
49   NaN   NaN   7.0  113.0   FSIQ4   9.0   

                                            TBI_notes  age_at_implant  \
0                                                 NaN       48.105026   
1                                                 NaN       48.105026   
2                                                 NaN       49.383743   
3                                                 NaN       49.383743   
4   motor vehicle accident due to loss of consciou...       39.390382   
5   motor vehicle accident due to loss of consciou...       39.390382   
6   In early 90s, motor vehicle accident resulting...       51.918931   
7                                                 NaN       20.553582   
8                                                 NaN       20.553582   
9                                                 NaN       30.801454   
10                                                NaN       30.801454   
11                                                NaN       36.581290   
12                                                NaN       54.509105   
13             physical assault sustained head injury       31.050603   
14             physical assault sustained head injury       31.050603   
15                                                NaN       47.727309   
16                                                NaN       48.655460   
17                    MVAs with loss of consciousness       38.678526   
18                                                NaN       24.753531   
19                                                NaN       32.668820   
20                                                NaN       36.759254   
21  birth injury, MVA  listed as ? in conference n...       24.923167   
22  birth injury, MVA  listed 

In [30]:
len(df_master)

1104

In [42]:
#Checking number of subjects included from different sites
sub_ar = df_master['subject'].unique()
res = sum(1 for i in sub_ar if "J" in i)
# sub_ar
res

63

 # Linear mixed effects model

In [32]:
# Compare models using a likelihood-ratio test :
from scipy import stats
import statsmodels.api as sm

def lrtest(llmin, llmax,df = 1):
    lr = 2 * (llmax - llmin)
    if llmax<llmin:
        raise ValueError('Check assumptions.')
    stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq, df)
    p = stats.chisqprob(lr, df) 
    return lr, p

The bellow statistics checks for experiment main effect (catFR/FR) on the total number of intrusions (not sig) and correct-recllas (catFR produce more correct recalls compared to FR)


In [57]:

vc = {'session': '0 + C(session)'} #the vc formula always varies within the groups parameter, so it will vary within subject 
# If we want to use list as another nested variable, we can use this:
# vc = {'session': '0 + C(session)', 'list': '0 + C(list):C(session)'}

## Total number of intrusions:
model = sm.MixedLM.from_formula('average_total_ints ~ 1',vc_formula=vc,  groups='subject', data=df_master) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# Experiment main effect
model = sm.MixedLM.from_formula('average_total_ints ~ C(experiment)',vc_formula=vc, groups='subject', data=df_master) 
result=model.fit(reml=False, method='nm')
llf_exp =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of experiment on total intrusions')
lr, p = lrtest(llf_empty, llf_exp, df = 1) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

## Total number of correct-recalls:
model = sm.MixedLM.from_formula('average_correct_recalls ~ 1',vc_formula=vc,  groups='subject', data=df_master) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# Experiment main effect
model = sm.MixedLM.from_formula('average_correct_recalls ~ C(experiment)',vc_formula=vc, groups='subject', data=df_master) 
result=model.fit(reml=False, method='nm')
llf_exp =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects of experiment on correct-recalls
print('Main effect of experiment on correct-recalls')
lr, p = lrtest(llf_empty, llf_exp, df = 1) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))



IndexError: index 1103 is out of bounds for axis 0 with size 1103

In [58]:
df_master.columns

Index(['BAI', 'BDI', 'ELIs', 'IQ', 'IQ_type', 'PLIs', 'TBI_notes',
       'age_at_implant', 'age_of_seizure_onset', 'all recalls', 'anx_type',
       'anxiety', 'average_ELI', 'average_PLI', 'average_correct_recalls',
       'average_total_ints', 'average_total_recalls', 'binary_TBI',
       'binary_prior_neuro', 'correct recalls', 'date_of_neuropsych_report',
       'dep_type', 'depression', 'experiment', 'gender', 'implant_date',
       'lists', 'prior_neurosurgical_procedures', 'reimplant_date', 'session',
       'site', 'subject', 'total ints', 'years_of_education'],
      dtype='object')

Let's start with a simple model that predicts intrusions/correct-reclls based on BAI+BDI scores (without other factors)

(We won't really report these results...)

In [59]:
## Total number of intrusions:

df_BAI_BDI_reduced=df_master.dropna(axis='index',subset=['BDI','BAI'])

model = sm.MixedLM.from_formula('average_total_ints ~ 1',vc_formula=vc,  groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# BDI+BAI variables - main effect
model = sm.MixedLM.from_formula('average_total_ints ~ BDI+BAI',vc_formula=vc, groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of BAI+BDI on total intrusions')
lr, p = lrtest(llf_empty, llf_demog, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


## Total number of correct-recalls:

model = sm.MixedLM.from_formula('average_correct_recalls ~ 1',vc_formula=vc,  groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# BDI+BAI variables - main effect
model = sm.MixedLM.from_formula('average_correct_recalls ~ BDI+BAI',vc_formula=vc, groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of BAI+BDI on total correct recalls:')
lr, p = lrtest(llf_empty, llf_demog, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

             Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_total_ints
No. Observations: 401     Method:             ML                
No. Groups:       120     Scale:              0.2366            
Min. group size:  1       Log-Likelihood:     -485.6048         
Max. group size:  10      Converged:          Yes               
Mean group size:  3.3                                           
------------------------------------------------------------------
               Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
------------------------------------------------------------------
Intercept      1.002      0.045   22.054   0.000    0.913    1.091
session Var    0.488      0.285                                   

             Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_total_ints
No. Observations: 401     Method:             ML                
No. Groups:       120     Scale:          

These results indicate that:
Higher BAI scores are associated with both more intrusions and more correct-reclls (suggesting that they have a more liberal threshorld for recalling words...?)
Higher BDI scores are associated with less correct recalls

Overall, BAI and BDI scores are significant predictors of the number of correct-recalls (trending significance for the number of intrusions)

We can also predict PLIs/ELIs based on BDI+BAI score:

In [60]:
##  number of PL intrusions:

df_BAI_BDI_reduced=df_master.dropna(axis='index',subset=['BDI','BAI'])

model = sm.MixedLM.from_formula('average_PLI ~ 1',vc_formula=vc,  groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# BDI+BAI variables - main effect
model = sm.MixedLM.from_formula('average_PLI ~ BAI+BDI',vc_formula=vc, groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_recall =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of BAI+BDI on average_PLI')
lr, p = lrtest(llf_empty, llf_recall, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

##  number of EL intrusions:

model = sm.MixedLM.from_formula('average_ELI ~ 1',vc_formula=vc,  groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# BDI+BAI variables - main effect
model = sm.MixedLM.from_formula('average_ELI ~ BAI+BDI',vc_formula=vc, groups='subject', data=df_BAI_BDI_reduced) 
result=model.fit(reml=False, method='nm')
llf_recall =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of BAI+BDI on average_ELI')
lr, p = lrtest(llf_empty, llf_recall, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


          Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_PLI
No. Observations: 401     Method:             ML         
No. Groups:       120     Scale:              0.0918     
Min. group size:  1       Log-Likelihood:     -267.8068  
Max. group size:  10      Converged:          Yes        
Mean group size:  3.3                                    
---------------------------------------------------------
                Coef. Std.Err.   z    P>|z| [0.025 0.975]
---------------------------------------------------------
Intercept       0.517    0.026 19.868 0.000  0.466  0.568
session Var     0.148    0.154                           

          Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_PLI
No. Observations: 401     Method:             ML         
No. Groups:       120     Scale:              0.0939     
Min. group size:  1       Log-Likelihood:     -258.2463  
Max. group size:  10      Converg

These results suggest that:
    Higher BAI scores are associated specifically with more PLIs (and not ELIs)

Here I predict the rate of intrusions/correct-recalls, based on: sex, prior resection, years of education, age at seizure onset, age of the patient, and general intelligence score (We can change these factors, but I chose here the ones that seemed the most relevant).

I then compare this model to a model that also includes BDI+BAI scores, to see whether those contribute to intrusions prediction/correct-recalls beyond these factors.

Note: the model cannot accept nan values. For each variable that has nan, we are loosing the whole subject, so for these variables let's try and maximize the data we have available. Currently only n=148 have the full data (with all demographic variables)

In [75]:
# demographic factors main effect on the total number of intrusions (not sig) and correct-recllas (catFR produce more correct recalls compared to FR)

list_demographics_columns=['years_of_education','age_at_implant','IQ','anxiety','depression'] # other anxiety and depression scores need to be normalized


# Total number of intrusions
# remove nan's from df:
df_master_reduced=df_master.dropna(axis='index',subset=list_demographics_columns)

## Total number of intrusions:
model = sm.MixedLM.from_formula('average_total_ints ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_ints ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on total intrusions')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


# demographis variables + BDI+BAI - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_ints ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for BDI/BAI
print('Main effect of BAI+BDI on total intrusions, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


## Total number of correct-recalls:
model = sm.MixedLM.from_formula('average_correct_recalls ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_correct_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on number of correct-recalls')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


# demographis variables + BDI+BAI - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_correct_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for BDI/BAI
print('Main effect of BAI+BDI on correct-recalls, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

             Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_total_ints
No. Observations: 305     Method:             ML                
No. Groups:       94      Scale:              0.2472            
Min. group size:  1       Log-Likelihood:     -376.5362         
Max. group size:  10      Converged:          Yes               
Mean group size:  3.2                                           
------------------------------------------------------------------
               Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
------------------------------------------------------------------
Intercept      0.966      0.054   17.817   0.000    0.859    1.072
session Var    0.523      0.323                                   

                 Mixed Linear Model Regression Results
Model:               MixedLM   Dependent Variable:   average_total_ints
No. Observations:    305       Method:               ML                
No. Groups:          94 

/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


                 Mixed Linear Model Regression Results
Model:              MixedLM Dependent Variable: average_correct_recalls
No. Observations:   305     Method:             ML                     
No. Groups:         94      Scale:              1.9841                 
Min. group size:    1       Log-Likelihood:     -537.2606              
Max. group size:    10      Converged:          Yes                    
Mean group size:    3.2                                                
-----------------------------------------------------------------------
                             Coef.  Std.Err.   z    P>|z| [0.025 0.975]
-----------------------------------------------------------------------
Intercept                    -0.421    0.725 -0.580 0.562 -1.842  1.001
C(gender)[T.Male]            -0.150    0.167 -0.897 0.369 -0.477  0.177
C(binary_prior_neuro)[T.1.0] -0.198    0.192 -1.032 0.302 -0.574  0.178
years_of_education            0.155    0.048  3.221 0.001  0.061  0.249
age_at_im

/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


In [74]:
select_subs = list(df_master_reduced['subject'].unique())
select_subs
with open("select_subs.txt", "w") as output:
    output.write(str(select_subs))

In [65]:
df_master_reduced=df_master.dropna(axis='index',subset=list_demographics_columns)

## Total number of intrusions:
model = sm.MixedLM.from_formula('average_PLI ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_PLI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on PLIs')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


# demographis variables + BDI+BAI - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_PLI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for BDI/BAI
print('Main effect of BAI+BDI on PLIs, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

## Total number of ELIs:
model = sm.MixedLM.from_formula('average_ELI ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_ELI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on ELIs')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


# demographis variables + BDI+BAI - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_ELI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for BDI/BAI
print('Main effect of BAI+BDI on ELIs, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

          Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_PLI
No. Observations: 305     Method:             ML         
No. Groups:       94      Scale:              0.0903     
Min. group size:  1       Log-Likelihood:     -183.1766  
Max. group size:  10      Converged:          Yes        
Mean group size:  3.2                                    
---------------------------------------------------------
                Coef. Std.Err.   z    P>|z| [0.025 0.975]
---------------------------------------------------------
Intercept       0.474    0.028 16.950 0.000  0.419  0.528
session Var     0.118    0.137                           

                 Mixed Linear Model Regression Results
Model:                  MixedLM     Dependent Variable:     average_PLI
No. Observations:       305         Method:                 ML         
No. Groups:             94          Scale:                  0.0920     
Min. group size:        1           Log-Li

In [69]:
#Evan's experimental mixed effects models:
list_demographics_columns=['years_of_education','age_at_implant','IQ','anxiety','depression'] 

# Total number of intrusions
# remove nan's from df:
df_master_reduced=df_master.dropna(axis='index',subset=list_demographics_columns)

## Total number of intrusions:
model = sm.MixedLM.from_formula('average_total_ints ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_ints ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on total intrusions')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + depression - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_ints ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for depression
print('Main effect of depression on total intrusions, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + anxiety - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_ints ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for anxiety
print('Main effect of anxiety on total intrusions, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

             Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_total_ints
No. Observations: 305     Method:             ML                
No. Groups:       94      Scale:              0.2472            
Min. group size:  1       Log-Likelihood:     -376.5362         
Max. group size:  10      Converged:          Yes               
Mean group size:  3.2                                           
------------------------------------------------------------------
               Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
------------------------------------------------------------------
Intercept      0.966      0.054   17.817   0.000    0.859    1.072
session Var    0.523      0.323                                   

                 Mixed Linear Model Regression Results
Model:               MixedLM   Dependent Variable:   average_total_ints
No. Observations:    305       Method:               ML                
No. Groups:          94 

In [77]:
#Check effect of depression and anxiety separately on PLIs
list_demographics_columns=['years_of_education','age_at_implant','IQ','anxiety','depression'] 

# Total number of intrusions
# remove nan's from df:
df_master_reduced=df_master.dropna(axis='index',subset=list_demographics_columns)

## Total number of intrusions:
model = sm.MixedLM.from_formula('average_PLI ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_PLI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on PLIs')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + depression - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_PLI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for depression
print('Main effect of depression on PLIs, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + anxiety - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_PLI ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for anxiety
print('Main effect of anxiety on PLIs, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

          Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_PLI
No. Observations: 305     Method:             ML         
No. Groups:       94      Scale:              0.0903     
Min. group size:  1       Log-Likelihood:     -183.1766  
Max. group size:  10      Converged:          Yes        
Mean group size:  3.2                                    
---------------------------------------------------------
                Coef. Std.Err.   z    P>|z| [0.025 0.975]
---------------------------------------------------------
Intercept       0.474    0.028 16.950 0.000  0.419  0.528
session Var     0.118    0.137                           

                 Mixed Linear Model Regression Results
Model:                  MixedLM     Dependent Variable:     average_PLI
No. Observations:       305         Method:                 ML         
No. Groups:             94          Scale:                  0.0920     
Min. group size:        1           Log-Li

In [70]:
###Check effect of anxiety and depression on total recalls
# remove nan's from df:
df_master_reduced=df_master.dropna(axis='index',subset=list_demographics_columns)

## Total number of recalls:
model = sm.MixedLM.from_formula('average_total_recalls ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on total recalls')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographic variables + depression - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for depression
print('Main effect of depression on total recalls, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + anxiety - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for anxiety
print('Main effect of anxiety on total recalls, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

# demographis variables + depression + anxiety - main effect ()
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_total_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for anxiety and depression
print('Main effect of anxiety and depression on total recalls, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))

               Mixed Linear Model Regression Results
Model:            MixedLM Dependent Variable: average_total_recalls
No. Observations: 305     Method:             ML                   
No. Groups:       94      Scale:              2.3058               
Min. group size:  1       Log-Likelihood:     -572.8285            
Max. group size:  10      Converged:          Yes                  
Mean group size:  3.2                                              
---------------------------------------------------------------------
                  Coef.   Std.Err.     z      P>|z|   [0.025   0.975]
---------------------------------------------------------------------
Intercept         4.734      0.092   51.465   0.000    4.554    4.914
session Var       0.202      0.240                                   



/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


                 Mixed Linear Model Regression Results
Model:              MixedLM  Dependent Variable:  average_total_recalls
No. Observations:   305      Method:              ML                   
No. Groups:         94       Scale:               2.1229               
Min. group size:    1        Log-Likelihood:      -547.5779            
Max. group size:    10       Converged:           Yes                  
Mean group size:    3.2                                                
-----------------------------------------------------------------------
                             Coef.  Std.Err.   z    P>|z| [0.025 0.975]
-----------------------------------------------------------------------
Intercept                     0.241    0.751  0.321 0.748 -1.230  1.712
C(gender)[T.Male]            -0.023    0.173 -0.131 0.896 -0.361  0.316
C(binary_prior_neuro)[T.1.0]  0.046    0.199  0.232 0.817 -0.343  0.435
years_of_education            0.150    0.050  3.022 0.003  0.053  0.247
age_at_im

/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


                 Mixed Linear Model Regression Results
Model:              MixedLM  Dependent Variable:  average_total_recalls
No. Observations:   305      Method:              ML                   
No. Groups:         94       Scale:               2.1076               
Min. group size:    1        Log-Likelihood:      -546.4724            
Max. group size:    10       Converged:           Yes                  
Mean group size:    3.2                                                
-----------------------------------------------------------------------
                             Coef.  Std.Err.   z    P>|z| [0.025 0.975]
-----------------------------------------------------------------------
Intercept                     0.248    0.748  0.332 0.740 -1.217  1.714
C(gender)[T.Male]            -0.073    0.175 -0.416 0.678 -0.416  0.270
C(binary_prior_neuro)[T.1.0]  0.060    0.198  0.304 0.761 -0.328  0.449
years_of_education            0.136    0.050  2.710 0.007  0.038  0.234
age_at_im

/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


                 Mixed Linear Model Regression Results
Model:              MixedLM  Dependent Variable:  average_total_recalls
No. Observations:   305      Method:              ML                   
No. Groups:         94       Scale:               2.1228               
Min. group size:    1        Log-Likelihood:      -547.5716            
Max. group size:    10       Converged:           Yes                  
Mean group size:    3.2                                                
-----------------------------------------------------------------------
                             Coef.  Std.Err.   z    P>|z| [0.025 0.975]
-----------------------------------------------------------------------
Intercept                     0.237    0.751  0.316 0.752 -1.235  1.710
C(gender)[T.Male]            -0.020    0.174 -0.114 0.909 -0.362  0.322
C(binary_prior_neuro)[T.1.0]  0.043    0.201  0.215 0.830 -0.350  0.436
years_of_education            0.150    0.050  3.020 0.003  0.052  0.247
age_at_im

/home1/evsnyder/.conda/envs/cmlreader0/lib/python3.7/site-packages/statsmodels/regression/mixed_linear_model.py:2149: ConvergenceWarning: The MLE may be on the boundary of the parameter space.
  warnings.warn(msg, ConvergenceWarning)


In [ ]:
## Total number of correct-recalls:
model = sm.MixedLM.from_formula('average_correct_recalls ~ 1',vc_formula=vc,  groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_empty =result.llf #log likelihood
print(result.summary())

# demographis variables - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_correct_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for main effects
print('Main effect of demographics on number of correct-recalls')
lr, p = lrtest(llf_empty, llf_demog, df = 5) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))


# demographis variables + BDI+BAI - main effect
# binary_prior_neuro is 'True' when prior neuro procedure was NOT present, 'False' when any king of procedure WAS present 
model = sm.MixedLM.from_formula('average_correct_recalls ~ C(gender)+C(binary_prior_neuro)+ years_of_education + age_at_implant + IQ + anxiety + depression',vc_formula=vc, groups='subject', data=df_master_reduced) 
result=model.fit(reml=False, method='nm')
llf_demog_Beck =result.llf #log likelihood
print(result.summary())

# Likelihood ratio tests for BDI/BAI
print('Main effect of BAI+BDI on correct-recalls, beyond other demographic variables:')
lr, p = lrtest(llf_demog, llf_demog_Beck, df = 2) # Degrees of freedom = number of regressors (i.e. columns in summary table) added when including the effect of interest, for categorical variable with two groups this is df = 1
print('LR test, p value: {:.4f}, {:.30f}'.format(lr, p))